### ISSUE TITLE
memory leak with WebEye.Controls.Wpf.WebCameraId

### ISSUE THREAD

**Issue by bouazaoui** [jacobbo/WebEye#31]
there is a memory leak with WebEye.Controls.Wpf.WebCameraId , i have taken the last source today.

However there are less memory leak with the Dispatcher.ShutdownStarted event dependency
thanks
https://snag.gy/w6LG2r.jpg

**Comment by jacobbo** [comment]
Hi, could you provide more details, please?

You can do the following test: create a child window with the control on it, then close the child window, do a GC.Collect() call and then count how many instances are left. I expect it to be zero.

**Comment by bouazaoui** [comment]
OK i will retry this on monday.
I have a Window with webcamera control instance , i closed it , webcameraid instance incrémented...
Thanks.
Best regards

⁣Envoyé par BlueMail ​

Le 15 sept. 2017 23:53, à 23:53, Alexander Iacobciuc <notifications@github.com> a écrit:

**Comment by Dmitry31** [comment]
Don't know if it is linked but when i switch multiple times from a page with WebCameraControl each time the instance is not disposed. It is caused by 

`parent.Closed += HandleWindowClosed;`

This reference will be kept by the window and will only be broken when window close event rises.
Ideally there should be both Loaded and Unloded event handlers. The first will tell the control to listen to the window close event the later to stop listening.

And in this way the following lines are usless:

`// https://stackoverflow.com/questions/3421303/loaded-event-of-a-wpf-user-control-fire-two-times                Loaded -= HandleControlLoaded;`

By the way handling Loaded event only once may cause errors in some scenarios like MVC approche where some control explicitly detaches/reattaches the WebCameraControl to the visual tree.

**Comment by jacobbo** [comment]
The idea here is to dispose the control only when a parent window is closed, in other words, their lifetimes should be the same. If the page in your example has not been closed (i.e. disposed), then we do not want to dispose the control.

**Comment by Dmitry31** [comment]
Well actually the lifetime of WPF control should not be the lifetime of the window! The control is not disposed when it is completly detached as it is stillreferenced by the Close event of the window. Lets see this most basic VM-first exemple 

XAML:
```
<Window x:Class="TestingWPF.MainWindow"
        xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:local="clr-namespace:TestingWPF"
        xmlns:wec="clr-namespace:WebEye.Controls.Wpf;assembly=WebEye.Controls.Wpf.WebCameraControl"
        Title="MainWindow" Height="350" Width="525">

    <Window.Resources>
        <DataTemplate DataType="{x:Type local:VM1}">
            <Rectangle Fill="Green" Margin="10"/>
        </DataTemplate>
        <DataTemplate DataType="{x:Type local:VM2}">
            <wec:WebCameraControl/>
        </DataTemplate>
    </Window.Resources>
    
    <Grid Background="#333333">
        <Grid.RowDefinitions>
            <RowDefinition Height="*"/>
            <RowDefinition Height="Auto"/>
        </Grid.RowDefinitions>

        <ContentControl x:Name="MainContainer">
            
        </ContentControl>

        <Button Content="Switch" Click="OnClick" Grid.Row="1"/>
    </Grid>
</Window>
```
Code behind:

```
namespace TestingWPF
{
    public class VM1 { }
    public class VM2 { }

    public partial class MainWindow : Window
    {
        private VM1 vm1 = new VM1();
        private VM2 vm2 = new VM2();

        public MainWindow()
        {
            InitializeComponent();
        }
        
        private void OnClick(object sender, RoutedEventArgs e)
        {
            MainContainer.Content = (MainContainer.Content == vm1) ? (object)vm2 : (object)vm1;
        }
    }
}
```
All it does is to switch the content ViewModel of the content control. Each VM has it's own data template that is instanciated each time the VM is set as content. Meaning that each time vm2 is the content of the MainContainer a new instance of WebCameraControl is created.

This exemple is pretty useless I didn't even start the capture. However it illustrates how basic WPF app would work. And what should happen here is that each time the vm1 is set and WebCameraControl is not in the visual tree any more the control should be disposed but it is not because auf the Window.Closed event handler in the control.

This should be handled in one of two ways:
1) Use Load to initialize everything and Unload to dispose everything this way everything will be disposed regardless of the window was closed or not. However in a scenario where the control is manually reattached to the visual tree (for exemple if instead of vm2 i used an instance of WebCameraControl) all the graph would be recreated from scratch eventually with performance issues.

2) Use Load to instanciate the DirectShow stuff but only if it was not yet done before. Use Unload to stop capture if it is still running. Use the destructor ~WebCameraControl to completly dispose all DirectShow part.

By the way if you run this exemple an you add a destructor ~WebCameraControl it will only be run when the window closes and as many times as the control was instanciated.

I used JetBrains dotMemory and it confirms that it's only the matter of Window.Closing event.

![image](https://user-images.githubusercontent.com/8328042/42992871-27fb2c8a-8c0a-11e8-83d0-6124b7d7a5ea.png)

**Comment by jacobbo** [comment]
Did you try your option #2? Does it resolve the memory leak? There is the VideoWindow (HwndHost) instance and I'm not sure if it's safe to dispose it in destructor.

**Comment by Dmitry31** [comment]
I tried with this destructor:

       ```
~WebCameraControl()
        {
            Dispatcher.BeginInvoke(
                System.Windows.Threading.DispatcherPriority.Normal,
                    ((Action)(() =>
                    {
                        if (_proxy != null)
                        {
                            if (_currentCamera != null)
                            {
                                StopCapture();
                            }

                            _proxy.DestroyCaptureGraph();
                            _proxy.Dispose();
                        }

                        _videoWindow.Dispose();
                    }))
                );
        }
```
It seem's to work in both scenarios MVVM (the exemple I have posted in previous message) and MVC (the same but i replavec VM1 and VM2 with instances of Rectangle and WebCamControl). But some more extensive testing would be required. Note that the destructor is called by GC that may take some time to fire. Fortunately the objects in memory are just few KB.

The handlers for the Loaded/Unloaded events can be something like this:

        ```
        private bool _resumeCaptureOnLoad = false;
        private WebCameraId _currentCameraBackup = null;

        private void HandleControlLoaded(object sender, RoutedEventArgs e)
        {
            if (_resumeCaptureOnLoad)
            {
                StartCapture(_currentCameraBackup);
                _resumeCaptureOnLoad = false;
                _currentCameraBackup = null;
            }
        }

        private void HandleControlUnloaded(object sender, RoutedEventArgs e)
        {
            if (IsCapturing)
            {
                _currentCameraBackup = _currentCamera;
                StopCapture();
                _resumeCaptureOnLoad = true;
            }
        }
```
So it restarts capturing if it was capturing before Unload. But mayby it is not the best behavior.

**Comment by jacobbo** [comment]
Sorry, I'm still not convinced this is a better way to stop/dispose the control. The Unloaded event is not reliable - it may not be raised at all. The destructor runs at indeterminate point in time - this may cause errors during stop/disposal.

**Comment by Dmitry31** [comment]
Yes that's why the #1 option was first. It is the way WPF controls should perform.

**Comment by jacobbo** [comment]
The #1 suggests usage of the Unloaded event, which is not reliable - it may not be raised at all.

**Comment by Dmitry31** [comment]
The only case, I see, where Unloaded event would not be fired is if OnLoaded event was not fired. That can only happen if the control was never attached to a visual tree. If you create a WebCameraControl and start capture but never attach to a visual tree then yes it will never be called ... However Window.Closed event will never be received either as a the control won't raise its Loaded event if never attached to visual tree. StartCapture should probably not be possible on a non loaded control. Or it should be postponed until Loaded event is fired.
Do you see any case where Loaded is fired but not Unloaded ?

**Comment by jacobbo** [comment]
I don't really care much about the Loaded event as the control does all the initialisation when you first call its methods, but the Unloaded counterpart looks not very reliable and its behaviour depends on the container control. Here are some examples for tab control and expander:
https://social.msdn.microsoft.com/Forums/vstudio/en-US/ee9672ed-a991-4dc1-9c92-a015ca2d99d3/when-exactly-are-the-loaded-unloaded-events-raised?forum=wpf
